File: packages/access-controls/src/chats.ts
import {
  and,
  assetPermissions,
  assetTypeEnum,
  chats,
  collectionsToAssets,
  eq,
  getDb,
  identityTypeEnum,
  isNull,
  usersToOrganizations,
} from '@buster/database';
import { z } from 'zod';

// Input validation schema
const CanUserAccessChatSchema = z.object({
  userId: z.string().uuid(),
  chatId: z.string().uuid(),
});

export const canUserAccessChat = async ({
  userId,
  chatId,
}: {
  userId: string;
  chatId: string;
}): Promise<boolean> => {
  // Validate inputs
  const input = CanUserAccessChatSchema.parse({ userId, chatId });
  
  const db = getDb();

  // Run all permission checks concurrently for optimal performance
  const [directPermission, collectionPermission, chatInfo, userOrgs] = await Promise.all([
    // Check 1: Direct user permission on chat
    checkDirectChatPermission(db, input.userId, input.chatId),
    
    // Check 2: User permission through collections
    checkCollectionChatPermission(db, input.userId, input.chatId),
    
    // Check 3: Get chat info (creator & organization)
    getChatInfo(db, input.chatId),
    
    // Check 4: Get user's organizations and roles
    getUserOrganizations(db, input.userId),
  ]);

  // If chat doesn't exist or is deleted, deny access
  if (!chatInfo) {
    return false;
  }

  // Check 1: Direct permission exists
  if (directPermission) {
    return true;
  }

  // Check 2: Collection permission exists
  if (collectionPermission) {
    return true;
  }

  // Check 3: User is the creator
  if (chatInfo.createdBy === input.userId) {
    return true;
  }

  // Check 4: User is workspace_admin or data_admin in the chat's organization
  const isOrgAdmin = userOrgs.some(
    (org) =>
      org.organizationId === chatInfo.organizationId &&
      (org.role === 'workspace_admin' || org.role === 'data_admin')
  );

  return isOrgAdmin;
};

// Helper function to check direct chat permission
async function checkDirectChatPermission(
  db: ReturnType<typeof getDb>,
  userId: string,
  chatId: string
): Promise<boolean> {
  const result = await db
    .select({ id: assetPermissions.assetId })
    .from(assetPermissions)
    .where(
      and(
        eq(assetPermissions.assetId, chatId),
        eq(assetPermissions.assetType, 'chat'),
        eq(assetPermissions.identityId, userId),
        eq(assetPermissions.identityType, 'user'),
        isNull(assetPermissions.deletedAt)
      )
    )
    .limit(1);

  return result.length > 0;
}

// Helper function to check collection-based chat permission
async function checkCollectionChatPermission(
  db: ReturnType<typeof getDb>,
  userId: string,
  chatId: string
): Promise<boolean> {
  const result = await db
    .selectDistinct({ collectionId: collectionsToAssets.collectionId })
    .from(collectionsToAssets)
    .innerJoin(
      assetPermissions,
      and(
        eq(assetPermissions.assetId, collectionsToAssets.collectionId),
        eq(assetPermissions.assetType, 'collection'),
        eq(assetPermissions.identityId, userId),
        eq(assetPermissions.identityType, 'user'),
        isNull(assetPermissions.deletedAt)
      )
    )
    .where(
      and(
        eq(collectionsToAssets.assetId, chatId),
        eq(collectionsToAssets.assetType, 'chat'),
        isNull(collectionsToAssets.deletedAt)
      )
    )
    .limit(1);

  return result.length > 0;
}

// Helper function to get chat info (creator and organization)
async function getChatInfo(
  db: ReturnType<typeof getDb>,
  chatId: string
): Promise<{ createdBy: string; organizationId: string } | null> {
  const result = await db
    .select({
      createdBy: chats.createdBy,
      organizationId: chats.organizationId,
    })
    .from(chats)
    .where(and(eq(chats.id, chatId), isNull(chats.deletedAt)))
    .limit(1);

  return result[0] || null;
}

// Helper function to get user's organizations and roles
async function getUserOrganizations(
  db: ReturnType<typeof getDb>,
  userId: string
): Promise<Array<{ organizationId: string; role: string }>> {
  const result = await db
    .select({
      organizationId: usersToOrganizations.organizationId,
      role: usersToOrganizations.role,
    })
    .from(usersToOrganizations)
    .where(
      and(
        eq(usersToOrganizations.userId, userId),
        isNull(usersToOrganizations.deletedAt)
      )
    );

  return result;
}

File: apps/api/libs/database/src/helpers/chats.rs
use crate::enums::{AssetPermissionRole, AssetType};
use anyhow::Result;
use diesel::JoinOnDsl;
use diesel::{BoolExpressionMethods, ExpressionMethods, QueryDsl, Queryable};
use diesel_async::RunQueryDsl;
use tokio::try_join;
use uuid::Uuid;

use crate::models::{AssetPermission, Chat};
use crate::pool::get_pg_pool;
use crate::schema::{asset_permissions, chats, collections_to_assets};

/// Fetches a single chat by ID that hasn't been deleted
///
/// # Arguments
/// * `id` - The UUID of the chat to fetch
///
/// # Returns
/// * `Result<Option<Chat>>` - The chat if found and not deleted
pub async fn fetch_chat(id: &Uuid) -> Result<Option<Chat>> {
    let mut conn = get_pg_pool().get().await?;

    let result = match chats::table
        .filter(chats::id.eq(id))
        .filter(chats::deleted_at.is_null())
        .first::<Chat>(&mut conn)
        .await
    {
        Ok(result) => Some(result),
        Err(diesel::NotFound) => None,
        Err(e) => return Err(e.into()),
    };

    Ok(result)
}

/// Helper function to fetch a chat by ID
async fn fetch_chat_helper(id: &Uuid) -> Result<Option<Chat>> {
    let mut conn = get_pg_pool().get().await?;

    let result = match chats::table
        .filter(chats::id.eq(id))
        .filter(chats::deleted_at.is_null())
        .first::<Chat>(&mut conn)
        .await
    {
        Ok(result) => Some(result),
        Err(diesel::NotFound) => None,
        Err(e) => return Err(e.into()),
    };

    Ok(result)
}

/// Helper function to fetch permission for a chat
async fn fetch_chat_permission(id: &Uuid, user_id: &Uuid) -> Result<Option<AssetPermissionRole>> {
    let mut conn = get_pg_pool().get().await?;

    let permission = match asset_permissions::table
        .filter(asset_permissions::asset_id.eq(id))
        .filter(asset_permissions::asset_type.eq(AssetType::Chat))
        .filter(asset_permissions::identity_id.eq(user_id))
        .filter(asset_permissions::deleted_at.is_null())
        .first::<AssetPermission>(&mut conn)
        .await
    {
        Ok(result) => Some(result.role),
        Err(diesel::NotFound) => None,
        Err(e) => return Err(e.into()),
    };

    Ok(permission)
}

/// Helper function to fetch collection-based permissions for a chat
///
/// Checks if the user has access to the chat through collections
/// by finding collections that contain this chat and checking
/// if the user has permissions on those collections
async fn fetch_collection_permissions_for_chat(
    id: &Uuid,
    user_id: &Uuid,
) -> Result<Option<AssetPermissionRole>> {
    let mut conn = get_pg_pool().get().await?;

    // Find collections containing this chat
    // then join with asset_permissions to find user's permissions on those collections
    let permissions = asset_permissions::table
        .inner_join(
            collections_to_assets::table.on(asset_permissions::asset_id
                .eq(collections_to_assets::collection_id)
                .and(asset_permissions::asset_type.eq(AssetType::Collection))
                .and(collections_to_assets::asset_id.eq(id))
                .and(collections_to_assets::asset_type.eq(AssetType::Chat))
                .and(collections_to_assets::deleted_at.is_null())),
        )
        .filter(asset_permissions::identity_id.eq(user_id))
        .filter(asset_permissions::deleted_at.is_null())
        .select(asset_permissions::role)
        .load::<AssetPermissionRole>(&mut conn)
        .await?;

    // Return any collection-based permission
    if permissions.is_empty() {
        Ok(None)
    } else {
        // Just take the first one since any collection permission is sufficient
        Ok(permissions.into_iter().next())
    }
}

#[derive(Queryable)]
pub struct ChatWithPermission {
    pub chat: Chat,
    pub permission: Option<AssetPermissionRole>,
}

pub async fn fetch_chat_with_permission(
    id: &Uuid,
    user_id: &Uuid,
) -> Result<Option<ChatWithPermission>> {
    // Run all queries concurrently
    let (chat, direct_permission, collection_permission) = try_join!(
        fetch_chat_helper(id),
        fetch_chat_permission(id, user_id),
        fetch_collection_permissions_for_chat(id, user_id)
    )?;

    // If the chat doesn't exist, return None
    let chat = match chat {
        Some(c) => c,
        None => return Ok(None),
    };

    // If collection permission exists, use it; otherwise use direct permission
    let effective_permission = match collection_permission {
        Some(collection) => Some(collection),
        None => direct_permission,
    };

    Ok(Some(ChatWithPermission {
        chat,
        permission: effective_permission,
    }))
}

/// Fetches multiple chats with their permissions in a single operation
///
/// # Arguments
/// * `ids` - Vector of UUIDs of the chats to fetch
/// * `user_id` - UUID of the user making the request
///
/// # Returns
/// * `Result<Vec<ChatWithPermission>>` - The chats with their permissions
pub async fn fetch_chats_with_permissions(
    ids: &[Uuid],
    user_id: &Uuid,
) -> Result<Vec<ChatWithPermission>> {
    if ids.is_empty() {
        return Ok(Vec::new());
    }

    let mut conn = get_pg_pool().get().await?;

    // 1. Fetch all chats
    let chats_data = chats::table
        .filter(chats::id.eq_any(ids))
        .filter(chats::deleted_at.is_null())
        .load::<Chat>(&mut conn)
        .await?;

    if chats_data.is_empty() {
        return Ok(Vec::new());
    }

    // 2. Fetch direct permissions for these chats
    let direct_permissions = asset_permissions::table
        .filter(asset_permissions::asset_id.eq_any(ids))
        .filter(asset_permissions::asset_type.eq(AssetType::Chat))
        .filter(asset_permissions::identity_id.eq(user_id))
        .filter(asset_permissions::deleted_at.is_null())
        .select((asset_permissions::asset_id, asset_permissions::role))
        .load::<(Uuid, AssetPermissionRole)>(&mut conn)
        .await?;

    // 3. Fetch collection-based permissions for these chats
    let collection_permissions = asset_permissions::table
        .inner_join(
            collections_to_assets::table.on(asset_permissions::asset_id
                .eq(collections_to_assets::collection_id)
                .and(asset_permissions::asset_type.eq(AssetType::Collection))
                .and(collections_to_assets::asset_id.eq_any(ids))
                .and(collections_to_assets::asset_type.eq(AssetType::Chat))
                .and(collections_to_assets::deleted_at.is_null())),
        )
        .filter(asset_permissions::identity_id.eq(user_id))
        .filter(asset_permissions::deleted_at.is_null())
        .select((collections_to_assets::asset_id, asset_permissions::role))
        .load::<(Uuid, AssetPermissionRole)>(&mut conn)
        .await?;

    // Create maps for easier lookup
    let mut direct_permission_map = std::collections::HashMap::new();
    for (asset_id, role) in direct_permissions {
        direct_permission_map.insert(asset_id, role);
    }

    // Create map for collection permissions (just take first one for each asset)
    let mut collection_permission_map = std::collections::HashMap::new();
    for (asset_id, role) in collection_permissions {
        // Only insert if not already present (first one wins)
        collection_permission_map.entry(asset_id).or_insert(role);
    }

    // Create ChatWithPermission objects with effective permissions
    let result = chats_data
        .into_iter()
        .map(|chat| {
            let direct_permission = direct_permission_map.get(&chat.id).cloned();
            let collection_permission = collection_permission_map.get(&chat.id).cloned();

            // Use collection permission if it exists, otherwise use direct permission
            let effective_permission = match collection_permission {
                Some(collection) => Some(collection),
                None => direct_permission,
            };

            ChatWithPermission {
                chat,
                permission: effective_permission,
            }
        })
        .collect();

    Ok(result)
}
